### PR TITLE
Node-develop: create routes for working with notes

### PR DESCRIPTION
**Pull request by petrovskya**
1. Overview: Created routes for working with notes. Notes can be created/edited/deleted/viewed, corresponding to the following request methods POST/PUT/DELETE/GET.

2. Changes:

   Task 2.
     - Created the following routes:
     
           GET - /api/notes - returns an empty array.
           POST - /api/notes - returns the note object sent in the request.
           PUT - /api/notes/:id - returns the note object sent in the request.
           DELETE - /api/notes/:id - returns {success: true, id: " value :id " }.
           
        The note contains several fields:
        
            { id - identifier - required field when editing
              title - note title - required field, at least 3 characters long
              content - note content - required field, at least 3 characters long and no longer than 500 characters
              createdAt - date of note creation
              updatedAt - date of note update, can be empty
             }
             
    Task 3.
     - Added input validation For POST and PUT requests. 
     - Returned the corresponding error to the user in case of incorrect data.
     - Added a logger of all requests to our server.
    
    Task 4. 
    - Created an account to create a free mongo database.
    - Returned the corresponding error to the user in case of database connection error.
     - Added the logic for saving notes to the database, editing them, deleting them and returning them to the user for viewing. 
     - Added pagination, the ability to filter notes by the date of creation, search by name.

     - Used the Render platform to demonstrate the result.
     - Prepared Postman requests in the collection.
     
     Task 5.
     - Initialized strategies for registration and login.
     - Added authorization to all routes except greetings.
 
     Task 6. 
     - Described the environment variable, when starting the server with which the postfix Dev is added to each new note.
     - Described the environment variable, when starting the server with which each new note is added with the Test postfix.

3. How to test:
[View Demo](https://lab-node-js.onrender.com/api/notes)
[View Demo. Postman Collection](https://elements.getpostman.com/redirect?entityId=28216450-6be95032-8c27-42ee-8ab6-816975faee07&entityType=collection)

### CODE REVIEW

File: src/api/notes/notes.service.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+@Injectable()
+export class NotesService {
+  private readonly notes: CreateNoteDto[] = [];
+  getAllNotes(): [] {
+    return EMPTY_ARRAY;
+  }
+
+  createNote(createNoteDto: CreateNoteDto): CreateNoteDto {
+    this.notes.push(createNoteDto);
+    return createNoteDto;
```
Comment (irina-komissarova) [inline src/api/notes/notes.service.ts]: please add entities folder inside /notes and add schema for Note 
https://docs.nestjs.com/techniques/mongodb

Comment (irina-komissarova) [reply src/api/notes/notes.service.ts]: and then use this.subscriptionModel.create

File: src/api/notes/notes.service.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  updateNote(updateNoteDto: UpdateNoteDto, id: string): UpdateNoteDto {
+    if (!id) {
+      throw new HttpException(ERROR_TEXT.BAD_REQUEST, HttpStatus.BAD_REQUEST);
+    }
+    return updateNoteDto;
```
Comment (irina-komissarova) [inline src/api/notes/notes.service.ts]: same
use models

File: src/api/notes/notes.service.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  deleteNote(id: string) {
+    return SUCCESS_DELETE_TEXT(id);
+  }
```
Comment (irina-komissarova) [inline src/api/notes/notes.service.ts]: you should get note by id, then check if it's not deleted (add error for this case) and then delete note
try to use soft delete (when note is not deleted, but field isDeleted and deletedAt fields added)

**Review by irina-komissarova — CHANGES_REQUESTED** [review]

File: src/api/notes/dto/notes.dto.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { Transform, TransformFnParams } from 'class-transformer';
+import { IsNumber, IsOptional, IsString } from 'class-validator';
+
+export class NotesDto {
+  @IsOptional()
+  @IsNumber()
+  @Transform(({ value }: TransformFnParams) => parseInt(value, 10))
```
Comment (irina-komissarova) [inline src/api/notes/dto/notes.dto.ts]: please add a separate util for it and reuse here (also remember about magic numbers)

File: src/api/notes/notes.service.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    page = DEFAULT_PAGE_VALUE,
+    limit = DEFAULT_NOTES_LIMIT_VALUE,
+    title,
+    createdAt,
+  }: NotesDto): Promise<Note[]> {
+    const skippedNotes = getSkipValue(page, limit);
+
+    const filter: FilterQuery<Note> = {
+      isDeleted: false,
+      ...(title && { title: new RegExp(escapeRegExp(title), 'i') }),
```
Comment (irina-komissarova) [inline src/api/notes/notes.service.ts]: it's better to create a consant for regexp value, f.e const titleRegexp = new RegExp(escapeRegExp(title), 'i');

File: src/api/notes/notes.service.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  async updateNote(updateNoteDto: UpdateNoteDto, id: string): Promise<Note> {
+    try {
+      const updatedNote = await this.noteModel.findByIdAndUpdate(
+        { _id: id },
+        { ...updateNoteDto, updatedAt: Date.now() },
+        { new: true },
+      );
+      return updatedNote;
```
Comment (irina-komissarova) [inline src/api/notes/notes.service.ts:91]: please check formatting, there should be an empty line before return statement

File: src/config/constants/errors.ts [hunk 7]
```diff
@@ -1,3 +1,13 @@
 export const ERROR_TEXT = {
   NAME_NOT_FOUND: 'No name to greet. Name not found',
+  BAD_REQUEST: 'Bad request',
 };
+
+export const ERROR_VALIDATION_MESSAGES = {
+  NOT_STRING: 'Should be a string.',
+  NOTE_TITLE: 'Title should be should be of minimum 3 characters length.',
```
Comment (irina-komissarova) [inline src/config/constants/errors.ts]: please check error message, it looks like text duplicating

File: src/utils/getEndOfTheDay.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+const END_HOURS = 23;
+const END_MINUTES = 59;
+const END_SECONDS = 59;
+const END_MILLISECONDS = 999;
+
```
Comment (irina-komissarova) [inline src/utils/getEndOfTheDay.ts:7]: try to use dayjs lib instead, i guess it will be easier

**Review by irina-komissarova — CHANGES_REQUESTED** [review]

File: src/utils/getEndOfTheDay.ts (hunk 8 above)
Comment (petrovskya) [reply src/utils/getEndOfTheDay.ts:7]: I haven't found a solution with dayjs, that would make work with these constants easier.

File: src/api/notes/notes.controller.ts [hunk 9]
```diff
@@ -7,32 +7,38 @@ import {
...
   Put,
   Query,
+  UseGuards,
 } from '@nestjs/common';
 
 import { PARAMS, ROUTES, SUB_ROUTES } from 'config/constants';
 
 import { CreateNoteDto, NotesDto, UpdateNoteDto } from './dto';
 import { NotesService } from './notes.service';
+import { AccessTokenGuard } from 'guards';
```
Comment (irina-komissarova) [inline src/api/notes/notes.controller.ts]: !

File: src/auth/auth.service.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    const tokens = await this.getTokens(user._id.toHexString(), user.email);
+
+    await this.updateRefreshToken(user._id.toHexString(), tokens.refreshToken);
+
+    return tokens;
+  }
+
+  async logOut(userId: string) {
+    console.log(userId);
```
Comment (irina-komissarova) [inline src/auth/auth.service.ts]: !

**Review by irina-komissarova — APPROVED** [review]

**Pull request merged**